Inline assembler FPU support

peterhinch opened the pull request:
1. strto_uint_t()  
   Written because I couldn't find a standard library function which could be called without build errors.
2. vmrs()  
   Implemented as a zero argument function for simplicity to copy the FPU status into the ARM status register. I'll provide functionality to copy to other ARM core regsiters if thought worthwhile.
3. functions vcvt_f32_s32 and vcvt_s32_f32  
   Dots replaced with underscores because dots cause problems with interpreter.
   Opcodes correspond with code generated by the C compiler but differ from those in the architecture manual, notably in bit 12 of the top word. I suspect a typo in the manual as the codes in the manual cause a crash.

peterhinch committed:
Arithmetic and vneg implemented

peterhinch committed:
Zero args and two FPU instructions done

peterhinch committed:
Mov (both directions) done

peterhinch committed:
All instructions done

peterhinch committed:
Complete bar more testing

peterhinch committed:
Tests added

peterhinch committed:
Tests added

dpgeorge commented:
Looks nice!  I'll review it in detail later.

dpgeorge commented on py/emitinlinethumb.c:
```diff
@@ -172,6 +172,46 @@ STATIC const char *get_arg_str(mp_parse_node_t pn) {
+
+STATIC mp_uint_t get_arg_fpreg(emit_inline_asm_t *emit, const char *op, mp_parse_node_t pn) {
+    const char *reg_str = get_arg_str(pn);
+    char *end ;
```
Need to fix white space before ;.

dpgeorge commented on py/emitinlinethumb.c:
```diff
@@ -172,6 +172,46 @@ STATIC const char *get_arg_str(mp_parse_node_t pn) {
+    char *end ;
+    mp_uint_t regno ;
+    regno = 0;
+    if (reg_str[0] == 's'){
```
Need to add whitespace before {

dpgeorge commented on py/emitinlinethumb.c:
```diff
@@ -172,6 +172,46 @@ STATIC const char *get_arg_str(mp_parse_node_t pn) {
+    regno = 0;
+    if (reg_str[0] == 's'){
+        regno =strto_uint_t(reg_str+1, &end) ;
+        if (regno <= 31 && end > reg_str +1)
```
Needs curly braces.

dpgeorge commented on py/emitinlinethumb.c:
```diff
@@ -465,7 +507,8 @@ STATIC void emit_inline_thumb_op(emit_inline_asm_t *emit, qstr op, mp_uint_t n_a
         if (MP_PARSE_NODE_IS_ID(pn_args[1])) {
             // second arg is a register (or should be)
-            mp_uint_t op_code, op_code_hi;
+            mp_uint_t op_code, op_code_hi, vd, vm;
```
vd and vm should be declared below when they are needed.

dpgeorge commented on py/emitinlinethumb.c:
```diff
@@ -503,7 +579,7 @@ STATIC void emit_inline_thumb_op(emit_inline_asm_t *emit, qstr op, mp_uint_t n_a
         } else {
             // second arg is not a register
-            mp_uint_t op_code;
+            mp_uint_t op_code, op_code2;
```
op_code2 is not needed as a variable.

dpgeorge commented:
vmrs should conform as much as possible to the ARM asm standard.  This is so that when people read the asm code they understand straight away what it does.  So even if you only support certain args, those args still need to be passed.  It would be `vmrs(APSR_nzcv, FPSCR)` and you'd need to verify the args are what you expect.

peterhinch committed:
Requested changes implemented. vmrs now copies FPU status to any ARM core register.
getlbl compound instruction implemented to simplify accessing data created with the data() statement.
The latter is the last of my currently planned enhacements to the inline assembler.

peterhinch commented:
Changes requested now implemented. Commit notes provide details.

dpgeorge commented:
Please remove the `getlbl` pseudo instruction.  That needs to be a separate PR so we can discuss it :)

peterhinch committed:
getlbl removed.

peterhinch commented:
Now removed. Learning the ropes, I'm afraid :)

dpgeorge commented:
Thank you!  I wanted to make this feature a compile-time option (not all MCUs have a VFP unit) so I reworked the code a little and merged in 5008972fefdd0d8cad214d4c0a2fb3daea0ac3c8.
